agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,226 @@
+	}
+
+	// PORTS
+	portsResource, ok := resources["PORTS_TCP"]
+	if ok {
+		taskPortsSlice := portsResource.StringSetValue
+		// For each port in current resource object, check if it is already 'consumed'. This is
+		// done by maintaining a list of consumed ports, i.e. list of all ports reserved by tasks
+		// being accounted for by host resources manager.
+		for _, port := range taskPortsSlice {
+			for _, consumedPort := range h.consumedResource["PORTS_TCP"].StringSetValue {
+				// If port is already reserved by some other task, this 'resources' object can not be consumed
+				if *port == *consumedPort {
+					return false, nil
+				}
+			}
+		}
+	}
+
+	// PORTS_UDP
+	portsUDPResource, ok := resources["PORTS_UDP"]
+	if ok {
+		taskPortsUDPSlice := portsUDPResource.StringSetValue
+		// Same for UDP ports
+		for _, port := range taskPortsUDPSlice {
+			for _, consumedPort := range h.consumedResource["PORTS_UDP"].StringSetValue {
+				if *port == *consumedPort {
+					return false, nil
+				}
+			}
+		}
+	}
+	// GPU
+	gpuResouce, ok := resources["GPU"]
+	if ok {
+		if *(h.initialHostResource["GPU"].IntegerValue) < *(h.consumedResource["GPU"].IntegerValue)+*(gpuResouce.IntegerValue) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (h *HostResourceManager) release(taskArn string, resources map[string]*ecs.Resource) {
+	h.hostResourceManagerRWLock.Lock()
+	defer h.hostResourceManagerRWLock.Unlock()
+	defer logger.Debug("Consumed resources after task release call", logger.Fields{
+		"taskArn":   taskArn,
+		"CPU":       *h.consumedResource["CPU"].IntegerValue,
+		"MEMORY":    *h.consumedResource["MEMORY"].IntegerValue,
+		"PORTS_TCP": h.consumedResource["PORTS_TCP"].StringSetValue,
+		"PORTS_UDP": h.consumedResource["PORTS_UDP"].StringSetValue,
+		"GPU":       *h.consumedResource["GPU"].IntegerValue,
+	})
+	if h.taskConsumed[taskArn] {
+		// CPU
+		*h.consumedResource["CPU"].IntegerValue -= *resources["CPU"].IntegerValue
+
+		// MEM
+		*h.consumedResource["MEMORY"].IntegerValue -= *resources["MEMORY"].IntegerValue
+
+		// PORTS
+		portsResource, ok := resources["PORTS_TCP"]
+		if ok {
+			taskPortsSlice := portsResource.StringSetValue
+			// Start removing ports one by one
+			for _, port := range taskPortsSlice {
+				// Create a copy of ports slice, iterate and find the port index in this slice
+				// then remove the port, create a new slice and assign it back to consumedResource["PORTS_TCP"].StringSetValue
+				itrPortSlice := h.consumedResource["PORTS_TCP"].StringSetValue
+				idx := 0
+				for i, consumedPort := range itrPortSlice {
+					if *consumedPort == *port {
+						idx = i
+					}
+				}
+				newPortSlice := append(itrPortSlice[:idx], itrPortSlice[idx+1:]...)
+				h.consumedResource["PORTS_TCP"].StringSetValue = newPortSlice
+			}
+		}
+
+		// PORTS_UDP
+		portsResource, ok = resources["PORTS_UDP"]
+		if ok {
+			taskPortsSlice := portsResource.StringSetValue
+			for _, port := range taskPortsSlice {
```

Comment:
All consumed ports for a single task are going to be in a contiguous sub-slice in `consumedResource["PORTS_UDP"]`, since we add them together. Therefore, instead of removing ports one at a time, I think we can find the starting and ending index and remove all at once.

Reply:
I was thinking of implementing as a generic slice check, but can get this optimization in as well